### DISCUSSION TITLE
understanding slave part of ebusd message

### DISCUSSION THREAD

**Discussion by JGJ156** [john30/ebusd#1371] [Q&A] [answered]
Dear all, I have got an ebusd shield 5, rpi 5, node red mqtt  emoncms set up, all working fine. I know the ebus wiki.

Next step would be to understand message definition files fully, in order to see whether I can improve the files. The master part I get, certain bytes in the message form the key to look up a line in the message definition file. However I am  not able to get my head around the slave part of the message. For example, if columns in definition file ctvl2 are:

# type (r[1-9];w;u),circuit,name,[comment],[QQ],ZZ,PBSB,[ID],field1,part (m/s),datatypes/templates,divider/values,unit,comment,field2,part (m/s),datatypes/templates,divider/values,unit,comment,field3,part (m/s),datatypes/templates,divider/values,unit,comment

grab result: f115b52406020003001400   / 08010314000000a041 = 1: ctlv2 z1ActualHeatingRoomTempDesired.
08 is the number of bytes following ->fine
ebusd would generate a message line where field 1 is filled with 01?
ebusd would generate a message line where part is filled with 03?
ebusd would generate a message line where datatype is filled with 14?
hex 14 is dec 20, 14 would be the wanted room temperature. But that is the next column?

the definition file determines tempv as the datatype, however that is not an existing datatype, so how is the code then interpreted....

I hope somebody can get me on the right track. I could not find a description clear enough for me.

best regards

**Comment by chrizzzp** [comment, marked as answer]
@JGJ156 

If message definitions are already there in the CSV (as in your case) it's easy. It's a good idea to look at the message definition and the message defaults above, e.g.:

```
# ##### zone 1 #####,,,,,,,,,,,,,
*r,,,,,,B524,02000300,,,IGN:4,,,

...

r,,z1ActualHeatingRoomTempDesired,Actual Desired Heating Room Temperature Zone 1,,,,1400,,,tempv,,,(0 = heating inactive)
```

Here, the IGN:4 in the default definition means the first 4 bytes of the slave response (after the first byte indicating length of the response) are ignored.

In case of 

`f115b52406020003001400 / 08010314000000a041 = 1: ctlv2 z1ActualHeatingRoomTempDesired`

this means `01031400` is ignored and the actual bytes evaluated are: `0000a041`

The datatype `tempv` is a template translating to the datatype EXP (`_templates.csv`)

`tempv,EXP,,°C,temperature`

If you use

`ebusctl grab result decode all`

and check the output for datatype EXP or simply type 

`ebusctl decode "EXP" 0000a041`

you will see that the result is '20' (20°C, the actual desired room temperature for zone 1).

Of course, if the message definitions are still unknown decoding is a bit more tricky. `ebusctl grab result decode all`  and tracking changes of unknown values over time helps a lot to identify their contents.

As far as I know this repo https://github.com/jonesPD/ebusd-configuration has the most complete set of definitions for the `ctlv2.csv`. I hope it will be merged some day with the default ebusd-configuration...

**Reply by JGJ156** [reply]
Thank you Chriszzzp, that is very helpfull. Yes, this is the 'very easy' part.

- I have imported the relevant csv files into excel, to see the columns better. 
- Okay, the datatype is then actually a combination of IGN:4 and Tempv, see now in the wiki what is meant with regards to _Fields part of a default row_... . 
- I was looking for Tempv in the _templates file in the EN folder, but it is only there in the _templates file in the Vaillant folder.....  
- all clear wrt ebusctl
- Should I be able to see why 0000a041 is 20 ? allthough there is a description in the Wiki, I don't see how to convert it,  
-  I am using the Jones PD files already, however there is still plenty to sort out.

**Reply by chrizzzp** [reply]
> * Should I be able to see why 0000a041 is 20 ? allthough there is a description in the Wiki, I don't see how to convert it,

`ebusctl decode "EXP" 0000a041`

There's floating point math behind it... (https://skftechnicalsupport.zendesk.com/hc/en-us/articles/360039409753-Understanding-Floating-point-big-little-endian - I'm not sure if this is what you're looking for...)

**Comment by JGJ156** [comment]
Thanks a lot Chriszzzp. Clear, from the wiki example in how to's I was under the impression that a simple hex to dec conversion should give a hint what the number could be, However all data types should be tried to find potential meanings. The converter below I also found. Also btw I did not realise that the help function in ebusctl can also be for a specific command, this also clarifies a lot. https://gregstoll.com/~gregstoll/floattohex/
![image](https://github.com/user-attachments/assets/e04d9eee-0ab8-4499-8451-4d28bdcee725)
